### ISSUE TITLE
What is the complexity of training an RNN sequence model? I cant find a good way to implement it in my deep learning library.

### ISSUE THREAD

**Issue by faruknane** [microsoft/CNTK#3750]
Hi, I've been developing a new [deep learning library](https://github.com/faruknane/DeepLearningFramework) in .Net Core where the derivatives of the terms can be calculated automatically by the library. You can think the terms as matrices. It is possible to create math equations and layers (and minimize them with gradient decent). Like below:
```csharp
public Variable W, B;
      
public override Term CreateTerm(int time)
{
        Term x = PreviousLayer.GetTerm(time);
        return new Plus(new MatrixMultiply(W, x), new ExpandWithSame(B, 1, x.D2));
}
```
Terms are the smallest calculation unit in my library. Layers are built on terms. When you run `CreateTerm(int time)` function, it only connects the terms and creates a big math expression. After a bit of giving what my library looks like, let me tell you how the derivative calculation works in my library (because I don't know what the standarts are, I just created and coded the thing). 

I calculate derivatives with four dimensional array (ofc not 4 dimension in optimized version, its more easy to tell you this way). 

```csharp
Matrix A, B, C;
C = A * B
```
For every output `(C)` and input `(A or B)` we have an element `Derivatives[m, n, k, l]` which equals to partial derivative of `C[m, n]` with respect to `A[k, l]` or `B[k, l]`. So, I store derivatives like that. 

How I combine derivatives is that, let the two derivatives be `D1[m, n, k, l]`, `D2[k, l, x, y]`, I calculate as `D3[m, n, x, y] = D1[m, n, i, j] * D2[i, j, x, y] for every i and j`. So in matrix multp. this calculation costs me O(N^3) due to passing the zero elements in D2. I also keep n and m equals one meaning the output always is size of (1,1) (a loss function) it reduces the complexity as well.

Notice that combining two derivatives means `D3[m, n, x, y] = D1[m, n, i, j] * D2[i, j, x, y] = (dF1 / dF2) * (dF2 / dF3) = dF1 / dF3` Everything is OK for now.

In a math expression which we want to minimize, there might be an expression used more than once.
<img height = "400" src="https://user-images.githubusercontent.com/37745467/64421878-507f0f00-d0ab-11e9-9023-5a8ed1b0483b.png"></img>

when backpropagating, we calculate the derivative of Block3 from left and right branches of the root two times.  In my opinion we don't have to calculate derivatives of 3rd Block with respect to its lower elements two times. For first calculation of `dRes / dBlock3` from left side, we can calculate derivatives of dBlock3 with respect to its lower elements and calculate the derivative of Res with respect to Block3's lower elements. Then we can store them to use it in calculation `dRes / dBlock3` from right side. But this has some consequences that we might wanna avoid. We gain a huge performance of Res being always size of (1,1) when combining two derivatives. I just don't wanna store and combine Derivative of Block3, because it will have 4 full dimension where every element is needed. and when combining it `dRes / dBlock3 * dBlock3 / dLowerElementsof3`. `D1[1, 1, x, y] = D2[1, 1, k, l] * D3[k, l, x, y]` because `D3` has full and every element of it is needed this calculation costs O(N^4). So its in practice not a good solution. 

So If we can't benefit from storing, there is only remaining way which is calculating it twice. If it is twice, its OK and can be tolerated. However, in RNN case, do I have to look up all back hidden states (0, 1, ..., T - 1) to train the time T. 

<img height = "250" src="https://user-images.githubusercontent.com/37745467/64424035-a904db00-d0b0-11e9-989c-73b709281a01.png"></img>

So RNN architecture costs T^2 time complexity? Instead of T? How is it calculated in CNTK with high performance?

**Comment by faruknane** [comment]
I solved the issue!
